File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.java
import java.util.*;
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		T = Integer.parseInt(br.readLine());
		StringBuilder sb = new StringBuilder();
		
		for(int t=1; t<=T; t++) {
			N = Integer.parseInt(br.readLine());
			arr = new int[N+1];
			st = new StringTokenizer(br.readLine());
			for(int i = 1; i<=N; i++) {
				arr[i] = Integer.parseInt(st.nextToken());
			}
			
			visited = new boolean[N+1];
			finished = new boolean[N+1];
			cnt = 0;
			for(int i = 1; i<=N; i++) {
				if(visited[i])
					continue;
				
				dfs(i);
			}
			
			sb.append(N - cnt).append('\n');
		}
		System.out.println(sb);
	}
	static int cnt;
	static void dfs(int cur) {
		
		visited[cur] = true;
		
		int nxt = arr[cur];
		if(!visited[nxt]) {
			dfs(nxt);
		}else {
			if(!finished[nxt]) {
				cnt++;
				for(int t = arr[nxt]; t != nxt; t = arr[t])
					cnt++;
			}
			
		}
		
		finished[cur] = true;
		
	}
	
	static boolean visited[], finished[];
	
	static StringTokenizer st;
	static int T, N;
	static int arr[];
	
}

File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.py
import sys

# 파이썬의 기본 재귀 깊이 제한은 1000 정도입니다.
# N이 최대 100,000이므로, 재귀 깊이 제한을 충분히 늘려주어야 런타임 에러(RecursionError)가 발생하지 않습니다.
sys.setrecursionlimit(110000)

def dfs(cur):
    global cnt
    nxt = arr[cur]

    if not visited[nxt]:
        visited[nxt] = True
        dfs(nxt)
    else:
        if not finished[nxt]:
            idx = nxt
            while True:
                cnt += 1
                idx = arr[idx]
                if idx == nxt:
                    break
                    
    finished[cur] = True



cnt = 0
if __name__ == "__main__":
    T = int(input())
    ans = []
    for _ in range(T):
        # print('-' * 10)
        N = int(input())
        arr = [0] + list(map(int,input().split()))
        visited = [False for _ in range(N+1)]
        finished = [False for _ in range(N+1)]
        cnt = 0
        for i in range(1, N + 1):
            if not visited[i]:
                dfs(i)

        # print(visited)
        ans.append(str(N - cnt))

    print('\n'.join(ans))
